### PR TITLE
Support SSR

### PR DESCRIPTION
**Pull request by matthova**
This PR is an alternative to https://github.com/Leaflet/Leaflet/pull/9316
It doesn't enforce nullish DOM access via lint, but, it does work 😅👍

### CODE REVIEW

**Comment by matthova** [comment]
@IvanSanchez Unfortunately optional chaining does not work if the window is undefined. The current pattern I'm using is to check via typeof
```js
const pointer = typeof window === 'undefined' ? false : !!window.PpinterEvent;
```

**Comment by jonkoops** [comment]
To get back to the discussion under #9316.

> My use case, as a consumer, is that I love being able to install a package and not have to think about it or do any additional configuration.

Whilst I understand this concern and we want to make using Leaflet as user-friendly as possible, I am left wondering what the implications of this are? I am not sure about adding support for SSR when Leaflet is very much not made to ever be used in a server-side context.

What will this PR accomplish exactly? Will this cause Leaflet code be executed on the server? If so, this would yield additional overhead of code execution that does essentially nothing. I cannot imagine this will actually get as far as to render the map on the server and re-hydrate it on the client. In which case I might even consider it a feature that Leaflet doesn't work in such a context.

Could you provide a minimal example of how Leaflet would be integrated in a Remix application, what the roadblocks are you are running into and how this will resolve said problem without introducing unnecessary overhead?

**Comment by matthova** [comment]
@jonkoops I have a lot of thoughts about Remix, but the bigger picture is that this is a problem that would need a unique solution across all of the current fad of server-rendered UI frameworks, NextJS, SvelteKit, Remix, Nuxt, etc... (I swear some of these framework names are made up 😅 )

In all of these style of frameworks, there is a server that does a first pass of the UI, and because the server doesn't have global DOM variables, SSR support puts the onus on library devs to handle nullish global dom variables. It's not ideal, but SSR support removes a big painpoint for people using libraries in these scenarios.

I take the mental load for maintainers seriously, so I understand wanting to avoid any unnecessary features. Looking at the diff in this PR, I'm cautiously convinced this is minimal lift and minimal compute for a high impact use-case. That said, the final call falls on the library devs. Thank you for the work you do 🙏🏻

**Comment by Falke-Design** [comment]
> Looking at the diff in this PR, I'm cautiously convinced this is minimal lift and minimal compute for a high impact use-case.

I see it like this too. Over time we got a few request for ssr support and if this few changes are all that are needed, then I would add them. But I would still not publish that we support ssr. It are just a few adjustments that makes it a little bit easier without offical support.

Other SSR references:
* #7085
* #6332

File: src/core/Util.js
Comment (Falke-Design) [inline src/core/Util.js:159]: I think `L.Util.falseFn` would be better. https://leafletjs.com/reference.html#util-falsefn

**Review by Falke-Design — CHANGES_REQUESTED** [review]

File: src/core/Util.js
Comment (matthova) [reply src/core/Util.js]: done 👍🏻

**Review by mourner — COMMENTED** [review]
The changes looks good to me! Fairly minimal for enabling a rare but increasing common use case. The only thing I'd suggest is seeing if we can add some kind of a test so that we don't accidentally regress on this in future PRs (which seems easy).

**Comment by jonkoops** [comment]
I am still not convinced we should be landing this at all. My questions still remains, why execute Leaflet code on the server when it cannot actually do anything useful? What are the problems that will occur if we don't land this change?

From what I am seeing here we're simply introducing more variations in code branching, and taking on supporting various server-side run-times while we're at it. Which implicates that we now need to take this on as additional maintenance should these run-times have different interpretations or expediencies of how the code should be run.

For example, Deno [has a `window` object](https://deno.land/api@v1.42.1?s=Window), so the guard code written in this PR will not work. And who knows what other run-times will come up with?

>  I have a lot of thoughts about Remix, but the bigger picture is that this is a problem that would need a unique solution across all of the current fad of server-rendered UI frameworks, NextJS, SvelteKit, Remix, Nuxt, etc... 

I understand that different SSR frameworks have different ways of dealing with client-side only code, but they all have escape hatches that allow users to run code on the client only (see [Next.js](https://nextjs.org/docs/pages/building-your-application/optimizing/lazy-loading), [Remix](https://remix.run/docs/en/main/guides/constraints#browser-only-code-on-the-server) and [SvelteKit](https://kit.svelte.dev/docs/page-options#ssr) documentation). If a user is committing to use a framework, I think it's a valid expectancy they learn how to use it.

**Comment by mourner** [comment]
@jonkoops I'm not familiar with various nuances of SSR frameworks, but we had a similar situation in Mapbox GL JS — tons of requests from users to make it possible to load the bundle in a server-side environment for who knows what. Eventually we gave up and merged a similar PR and added a test which simply checked that `node mapbox-gl.js` doesn't throw any errors, and there were no issues or any kind of maintenance burden since. 

Requiring Leaflet in some server script might not be a use case that makes a lot of sense for us (without much experience in this kind of SSR projects), but people will keep requesting it, and IMO it's a small price to pay.

> why execute Leaflet code on the server when it cannot actually do anything useful?

Even if we don't consider actually rendering something on the server, they could at least use some utility functions Leaflet provides without resorting to another tool like Turf — geometric / geographical calculations, projections, etc.

**Comment by jonkoops** [comment]
Very well, if we have to land this I'll give this a proper pass. Let me take a look and put in a review.

**Review by jonkoops — CHANGES_REQUESTED** [review]
There are still some globals referred to that might cause issues in a server-side run-time, but they will only create issues if DOM related APIs are used on the server, or if a `Map` is initialized.

Not sure what we want to do with those, handling them on a server is not really possible, and the guard code would be hundreds of lines of code.

**Comment by IvanSanchez** [comment]
> I am still not convinced we should be landing this at all. My questions still remains, why execute Leaflet code on the server when it cannot actually do anything useful? What are the problems that will occur if we don't land this change?

@jonkoops I agree with the sentiment. All that this will do is waste some RAM in JS web servers, fail some checks, and produce nothing.

On the other hand, I think of this the same way I think of old MSIE compatibility code: we don't need it for the "normal" use case, it shouldn't be there, it's ugly but, but makes things work in some minority scenario.

File: src/core/Util.js
Comment (mourner) [reply src/core/Util.js:160]: @jonkoops I don't think we should be that thorough here — IMO we should do the minimum possible change that makes the bundle itself execute without errors, disregarding how it will be used in such environment. Putting too much guards just bloats the code for cases no one will run into.

**Comment by mourner** [comment]
> There are still some globals referred to that might cause issues in a server-side run-time, but they will only create issues if DOM related APIs are used on the server, or if a `Map` is initialized.

In summary, I think we shouldn't anticipate any possible needs on this front by doing premature changes — instead let's accept the most minimal changes possible, as they look harmless and should cover immediate bases. Users who use Leaflet in SSR contexts can always follow-up with small fixes later, which can be reviewed separately.

File: src/core/Util.js
Comment (jonkoops) [reply src/core/Util.js:160]: Fair, just want to prevent that in this case things become polymorphic. In this case `function` and `undefined`.

**Comment by jonkoops** [comment]
Ok, if the issues that would occur under Deno SSR are fixed this can be merged as far as I am concerned.

**Comment by Gugustinette** [comment]
Thanks for considering these changes ! 🥹 This is also a common discussion in packages such as [Nuxt Leaflet](https://github.com/Gugustinette/Nuxt-Leaflet) which I work on.

We do not exactly want to **_run_** Leaflet on the server, but just having it's instance **_loaded_** without throwing errors, so SSR environments can play with it (and most of the time, send it to the client on demand).
This will make Leaflet very easy to use in these magic and modern framework.

I can try a kind of canary version of Leaflet on my module before merging if you want.

**Comment by NilsBaumgartner1994** [comment]
PLease make this work :-)

**Comment by Falke-Design** [comment]
I accidentally closed this PR by adding the wanted changes ... I reopened it (#9385)

I added the changes from @jonkoops and made it work with Deno.

@Gugustinette can you please test the version of #9385?

**Comment by Falke-Design** [comment]
test for reopen PR

**Comment by Falke-Design** [comment]
@matthova if its possible for you to use the commits from #9385 and reopen this PR again, I'm happy to close my request again

**Pull request closed without merging**
